src/metabase/setup.clj

```diff
@@ -25,12 +26,14 @@
   ;;
   ;; TODO -- 95% sure we can just use [[setup-token]] directly now and not worry about manually fetching the env var
   ;; value or setting DB values and the like
-  (let [mb-setup-token (env/env :mb-setup-token)]
-    (or (when mb-setup-token (setting/set-value-of-type! :string :setup-token mb-setup-token))
-        (db/select-one-field :value Setting :key "setup-token")
-        (setting/set-value-of-type! :string :setup-token (str (UUID/randomUUID))))))
+  (or (when-let [mb-setup-token (env/env :mb-setup-token)]
+        (setting/set-value-of-type! :string :setup-token mb-setup-token))
+      (db/select-one-field :value Setting :key "setup-token")
+      (setting/set-value-of-type! :string :setup-token (str (UUID/randomUUID)))))
 
-(defn clear-token!
-  "Clear the setup token if it exists and reset it to `nil`."
-  []
-  (setting/set-value-of-type! :string :setup-token nil))
+(defsetting has-user-setup
+  "A value that is true iff the metabase instance has one or more users registered."
+  :visibility :public
+  :type       :boolean
+  :setter     :none
+  :getter     (fn [] (db/exists? User)))
```

Comment:
We can't ever become un-set-up after we're set up (we specifically disallow deleting the last User) and since this is a `:public` Setting that gets returned every time someone loads up the app I think it's worth optimizing out the application DB hit. One little hit like this is not a huge deal but they tend to add up over time if we don't try to be aggressive about eliminating them where we can.

I would do something like this:

```suggestion
(defsetting has-user-setup
  "A value that is true iff the metabase instance has one or more users registered."
  :visibility :public
  :type       :boolean
  :setter     :none
  ;; Once a User is created it's impossible for this to ever become falsey -- deleting the last User is disallowed.
  ;; After this returns true once the result is cached and it will continue to return true forever without any
  ;; additional DB hits.
  :getter     (let [user-exists? (atom false)]
                (fn []
                  (or @user-exists?
                      (reset! user-exists? (db/exists? User))))))
```

Reply:
I'm still a bit hazy on how the caching system works but would the cache not prevent this from hitting the db all the time?

Reply:
Caching is part of the default get-value-from-application-DB getters but for a Setting with a completely custom getter the cache is bypassed. Calling `(has-user-setup)` or `(setting/get :has-user-setup)` is effectively the same as just calling the `:getter` function here.

This is intentional, since custom getters are usually returning calculated values caching them wouldn't be desirable in most cases. For example because the cache is set up to flush when rows in the `setting` Table change in the app DB, it wouldn't pick up changes to the `User` table and could end up returning the wrong value if this was cached

Reply:
Makes sense. Thanks